Add `ppairDataBuiltin`; Make `Tracing` tests structure consistent with other tests

TotallyNotChase opened the pull request.

L-as commented on Plutarch/Builtin.hs:
```diff
@@ -28,8 +32,24 @@ import PlutusTx (Data)
+
+Uses 'PAsData' to preserve more information about the underlying 'PData'.
+-}
+ppairDataBuiltin :: Term s (PAsData a :--> PAsData b :--> PBuiltinPair (PAsData a) (PAsData b))
```
We probably want to fix the `IsData` instance for pairs then to make it use `PAsData`.

TotallyNotChase replied:
Not sure what you meant here. The `IsData` instance is just for `(Integer, [Data])` right? The 2 members aren't `Data` values so we can't use `PAsData` there.

L-as replied:
You are right.

L-as replied:
However, why do you need this then?

L-as commented on Plutarch/Builtin.hs:
```diff
@@ -39,6 +59,20 @@ ptailBuiltin = phoistAcyclic $ pforce $ punsafeBuiltin PLC.TailList
+This uses 'PAsData' to preserve more information about the underlying 'PData'.
+-}
+pnilPairDataBuiltin :: Term s (PBuiltinList (PBuiltinPair (PAsData a) (PAsData b)))
+pnilPairDataBuiltin = punsafeConstant . PLC.Some $ PLC.ValueOf (PLC.DefaultUniList $ PLC.DefaultUniPair PLC.DefaultUniData PLC.DefaultUniData) []
```
While it's easy to add these functions, I don't think we should unfortunately, since we will have to deprecate them very soon. Notably, we want a generic `pconstant` that works for all combinations of literals as I've mentioned before. I can look into designing this if you're having a hard time imagining what it should look llike.

TotallyNotChase replied:
Definitely want `pconstant` for all this stuff. Have that on my todo list a well. These here are really just meant to be alternatives that people can use in the meantime. As I have been myself in the validator rewrite.

L-as replied:
What is this function used for?

TotallyNotChase replied:
I often find myself needing a `nil`. On the validator rewrites, I mostly use builtin list of `Data` - where I needed a `pnilDataBuiltin` to implement `filterData :: Term s (PBuiltinList a :--> PBuiltinList a)`.

I haven't needed `pnilPairDataBuiltin` it - but we'll need it when working with plutus assoc maps. The `Map` data value, `[(Data, Data)]`. This will end up representing things like `Value` - and implementing utilities on that `[(Data, Data)]` may require a `pnilPairDataBuiltin`.

L-as replied:
I am currently leaning toward exploring #53 a bit more before deciding on this.

TotallyNotChase replied:
It'd be great to have #53 resolved so soon! No need to have specific `nil`s in that case. I'll remove these from this PR.

srid replied:
With #62 you get write this functions as:

```haskell
pnil :: forall k a (s :: k). PLC.Contains PLC.DefaultUni (PDefaultUniType a) => Term s (PBuiltinList a)
pnil = pconstant @(PBuiltinList a) []

pnilPairData :: forall k (s :: k). (Data, Data) -> Term s (PBuiltinPair PData PData)
pnilPairData = pconstant @(PBuiltinPair PData PData)
```

L-as replied:
Why do you need the type applications?

L-as reviewed (changes requested):
See comment. Thanks for the contribution!

L-as commented:
We have `pconstant` now.

TotallyNotChase commented:
> We have `pconstant` now.

Awesome! Removed the `nil` stuff. Also removed the `pconsBuiltin` in favor of having that land in #63 

I guess this PR just adds `ppairDataBuiltin` now lol. Also, I moved out `Plutarch.Spec.Tracing` and put it inside `Examples.Tracing` for consistency.

L-as reviewed (approved):
LGTM
